src/time_series.c

```diff
@@ -0,0 +1,201 @@
+#include <private/rdycoreimpl.h>
+#include <private/rdymemoryimpl.h>
+
+static PetscErrorCode InitBoundaryFluxes(RDy rdy) {
+  PetscFunctionBegin;
+
+  // Allocate per-boundary flux offsets
+  PetscCall(RDyAlloc(PetscInt, rdy->num_boundaries + 1, &(rdy->time_series.boundary_fluxes.offsets)));
+
+  // count the number of global boundary edges (excluding those for
+  // auto-generated boundary conditions and those not locally owned)
+  // and compute local flux offsets
+  PetscInt num_boundary_edges = 0;
+  for (PetscInt b = 0; b < rdy->num_boundaries; ++b) {
+    RDyCondition bc = rdy->boundary_conditions[b];
+    if (!bc.auto_generated) {
+      RDyBoundary *boundary = &rdy->boundaries[b];
+      for (PetscInt e = 0; e < boundary->num_edges; ++e) {
+        PetscInt edge_id = boundary->edge_ids[e];
+        PetscInt cell_id = rdy->mesh.edges.cell_ids[2 * edge_id];
+        if (rdy->mesh.cells.is_local[cell_id]) ++num_boundary_edges;
+      }
+    }
+    rdy->time_series.boundary_fluxes.offsets[b + 1] = 3 * num_boundary_edges;
```

Comment:
Let's get `3 (=ndof)` from PETSc 

```
  PetscInt ndof;
  PetscCall(VecGetBlockSize(rdy->X_local, &ndof));
```

Reply:
This is a little more complicated, because this `3` is actually hard-wired in various ways at the moment. Among them:

* The `rdy.time_series.boundary_fluxes` struct has three named fields that identify the boundary fluxes
* The `AccumulateBoundaryFluxes` function takes an array whose last dimension is `3` (the number of boundary fluxes), and changing that to a runtime parameter would be a little awkward (but doable)

We could check that `ndof == 3` above and leave a note in `rdycoreimpl.h` that the number of DOF must match the number of boundary fluxes. I don't really want to give the impression that everything will magically work when someone changes one of these numbers.

Reply:
Okay, let's leave `3` for now. When we will add the sediment dynamics based on @donghuix's https://github.com/RDycore/toy-problems/pull/22, we will automatically have to revisit the hard-wired value of `3` in multiple places.